pull/github.go

```diff
@@ -328,19 +338,40 @@ func (ghc *GitHubContext) Commits() ([]*Commit, error) {
 }
 
 func (ghc *GitHubContext) PushedAt(sha string) (time.Time, error) {
+	repoID := ghc.pr.BaseRepository.DatabaseID
+	if ghc.pushedAt == nil {
+		ghc.pushedAt = make(map[string]time.Time)
+	}
+
+	// When loading pushed dates, prefer the local cache in the context. This
+	// helps if we don't have a global cache and avoids the possibility that
+	// the entry in the global cache is evicted while we process a single
+	// request.
+	//
+	// If the local context cache does not have a value, try the global cache,
+	// if it exists. If that's also missing an entry, then we'll compute the
+	// value from the GitHub API.
+
 	if t, ok := ghc.pushedAt[sha]; ok {
 		return t, nil
 	}
+	if gc := ghc.globalCache; gc != nil {
+		if t, ok := gc.GetPushedAt(repoID, sha); ok {
+			ghc.pushedAt[sha] = t
+			return t, nil
+		}
+	}
 
 	pushedAt, err := ghc.loadPushedAt(sha)
 	if err != nil {
 		return time.Time{}, err
 	}
 
-	if ghc.pushedAt == nil {
-		ghc.pushedAt = make(map[string]time.Time)
-	}
 	ghc.pushedAt[sha] = pushedAt
+	if gc := ghc.globalCache; gc != nil {
+		gc.SetPushedAt(repoID, sha, pushedAt)
```

Comment:
I need to fix this so that it doesn't cache the zero time, since that indicates "no pushed date" and is not globally cacheable.

Note that we might end up with the same commit having two pushed times as a result: the first evaluation will use the evaluation time and then future evaluations will use the time of the first status. I'd like to cache the evaluation time instead, but still trying to figure out how to do that in a way that makes sense, since it interacts with the commit filtering implemented at a different level.